File: Go/src/ray3_64bit/raytrace.go
package main

import (
	"image"
	"image/color"
	"math"

	. "github.com/go-gl/mathgl/mgl64"
)

const divisions = 8
const maxDepth = 5

type PixelResult struct {
	x, y  int
	color Color
}

func renderToImage(scene Scene, i *image.RGBA) {
	b := i.Bounds()
	height := float64(b.Dy())
	width := float64(b.Dx())

	getPoint := func(x, y float64, camera Camera) Vec3 {
		recenterX := func(x float64) float64 { return (x - (width / 2.0)) / 2.0 / width }
		recenterY := func(y float64) float64 { return -(y - (height / 2.0)) / 2.0 / height }
		recenterRight := camera.right.Mul(recenterX(x))
		recenterUp := camera.up.Mul(recenterY(y))
		return recenterRight.Add(recenterUp).Add(camera.forward).Normalize()
	}

	// collector := make(chan PixelResult)

	EmitRay := func(x, y int) PixelResult {
		ray := Ray{scene.camera.position, getPoint(float64(x), float64(y), scene.camera)}
		c := traceRay(ray, scene, 0)
		return PixelResult{
			x, y,
			c,
		}
		// collector <- PixelResult{x, y, c}
	}

	for y := b.Min.Y; y < b.Max.Y; y++ {
		for x := b.Min.X; x < b.Max.X; x++ {
			// if x == 40 && y == 40 {
			// 	EnableLogging()
			// } else {
			// 	DisableLogging()
			// }

			res := EmitRay(x, y)
			c := res.color.toDrawingColor()

			i.SetRGBA(x, y, color.RGBA{uint8(c.r), uint8(c.g), uint8(c.b), 255})

			// ray = Ray{scene.camera.position, getPoint(float64(x), float64(y), scene.camera)}
			// Log("Scene.Camera: ", scene.camera)
			// Log("X:", x, " Y:", y)
			// Log("Ray: ", ray.start, ray.direction)
			// Log("Float color : ", c)
			// Log("UINT color : ", c2)
		}
	}
	// fmt.Println("goroutines started.")

	// // Pull from collecter
	// count := height * width
	// for ; count != 0; count-- {
	// 	res := <-collector
	// 	c := res.color.toDrawingColor()
	// 	i.SetRGBA(res.x, res.y, color.RGBA{uint8(c.r), uint8(c.g), uint8(c.b), 255})
	// }
}

func intersections(ray Ray, scene Scene) (i Intersection, hit bool) {
	var closest float64 = math.MaxFloat64
	for _, t := range scene.things {
		inter, h := t.intersect(ray)
		if h && inter.distance < closest {
			i = inter
			hit = true
			closest = inter.distance
		}
	}
	return
}

func testRay(ray Ray, scene Scene) (dist float64, hit bool) {
	isect, hit := intersections(ray, scene)
	if hit == true {
		return isect.distance, true
	}
	return 0, false
}

func traceRay(ray Ray, scene Scene, depth int) Color {
	isect, hit := intersections(ray, scene)
	if hit == true {
		return shade(isect, scene, depth)
	}
	return Background()
}

// Shading

func shade(isect Intersection, scene Scene, depth int) Color {
	d := isect.ray.direction
	pos := d.Mul(isect.distance).Add(isect.ray.start)
	normal := isect.thing.normal(pos)
	reflectDir := d.Sub(normal.Mul(normal.Dot(d)).Mul(2))
	naturalColor := Plus(Background(),
		NaturalColor(isect.thing, pos, normal, reflectDir, scene))
	reflectedColor := Gray()
	if depth <= maxDepth {
		reflectedColor = ReflectionColor(isect.thing, pos, normal, reflectDir, scene, depth)
	}
	return Plus(naturalColor, reflectedColor)
}

func ReflectionColor(thing Thing, position, normal, rd Vec3, scene Scene, depth int) Color {
	c := traceRay(Ray{position, rd}, scene, depth+1)
	return c.Scale(thing.Surface().reflect(position))
}

func NaturalColor(thing Thing, position, normal, rd Vec3, scene Scene) Color {
	addLight := func(color Color, light Light) Color {
		ldis := light.position.Sub(position)
		livec := ldis.Normalize()
		neatIsect, hit := testRay(Ray{position, livec}, scene)
		var isInShadow bool
		if hit && neatIsect <= ldis.Len() {
			isInShadow = true
		}
		if isInShadow {
			return color
		} else {
			illum := livec.Dot(normal)
			lcolor := DefaultColor()
			if illum > 0 {
				lcolor = light.color.Scale(illum)
			}
			specular := rd.Normalize().Dot(livec)
			scolor := DefaultColor()
			if specular > 0 {
				scolor = light.color.Scale(float64(math.Pow(float64(specular), float64(thing.Surface().Roughness()))))
			}
			return Plus(color, Plus(Times(thing.Surface().diffuse(position), lcolor),
				Times(thing.Surface().specular(position), scolor)))
		}
	}
	acc := DefaultColor()
	for _, l := range scene.lights {
		acc = addLight(acc, l)
	}
	return acc
}

File: Python/raytrace.py
import math
from collections import namedtuple
from functools import reduce 
from typing import Callable

import color
from color import Color
from vector import Vector
# from surface import shiny, checkerboard
from thing import Thing

from ray_types import Ray, Intersection, Scene, Light, Camera

MAX_DEPTH = 5

def intersections(ray: Ray, scene: Scene) -> Intersection:
    closest = math.inf
    closestInter = None
    for t in scene.things:
        inter = t.intersect(ray)
        if inter != None and inter.distance < closest:
            closestInter = inter
            closest = inter.distance
    return closestInter

def test_ray(ray: Ray, scene: Scene) -> float:
    isect = intersections(ray, scene)
    if isect:
        return isect.distance
    else:
        return None

def trace_ray(ray: Ray, scene: Scene, depth: int) -> Color:
    isect = intersections(ray, scene)
    if not isect:
        return color.background
    else:
        return shade(isect, scene, depth)

def shade(isect: Intersection, scene: Scene, depth: int) -> Color:
    d = isect.ray.dir
    pos = isect.ray.start + d * isect.distance
    normal = isect.thing.normal(pos)
    reflect_dir = d - 2 * (normal * (normal & d))
    natural_color = color.plus( color.background,  get_natural_color(isect.thing, pos, normal, reflect_dir, scene) )
    reflected_color = ( color.grey if depth >= MAX_DEPTH else
        get_reflection_color(isect.thing, pos, normal, reflect_dir, scene, depth) )
    return color.plus(natural_color, reflected_color)

def get_reflection_color(thing: Thing, pos:Vector, normal:Vector, rd:Vector, scene:Scene, depth: int) -> Color:
    return color.scale(thing.surface.reflect(pos), 
        trace_ray(Ray(start=pos, dir=rd), scene, depth + 1))

def get_natural_color(thing: Thing, pos:Vector, norm: Vector, rd: Vector, scene: Scene) -> Color:
    def add_light(c: Color, l: Light):
        # print('c=',c)
        ldis = l.pos - pos
        livec = ldis.normal()
        neatIsect = test_ray(Ray(start=pos, dir=livec), scene)
        isInShadow = False if not neatIsect else neatIsect <= ldis.magnitude()
        if isInShadow:
            return c
        else:
            illum = livec & norm
            lcolor = color.default_color if illum < 0 else color.scale(illum, l.color)
            specular = livec & rd.normal()
            scolor = color.default_color if specular < 0 else color.scale(specular**thing.surface.roughness, l.color)
            # print('[types] c:{} pos:{} lcolor:{} scolor:{}'.format(c,pos,lcolor, scolor))
            # inter1 = thing.surface.diffuse(pos) 
            # inter2 = thing.surface.specular(pos) 
            # print('[types] inter1:{} inter2:{}'.format(inter1, inter2))
            return color.plus(c, (color.times(thing.surface.diffuse(pos) ,lcolor) + color.times(thing.surface.specular(pos), scolor)) )

    res = reduce(add_light, iter(scene.lights), color.default_color)
    # print('res=',res)
    return res

def render_to_image(scene: Scene, width, height, pixelOutput: Callable[[int, int, Color], None] ):
    '''pixelOutput called with color dict (rgb)'''
    
    def get_point(x: float, y: float, camera:Camera):
        recenterX = lambda x: (x - (width / 2.0)) / 2.0 / width
        recenterY = lambda y: -(y - (height / 2.0)) / 2.0 / height
        v = camera.forward + ( recenterX(x) * camera.right + recenterY(y) * camera.up )
        # print('v=', v)
        return Vector.normal(v)

    for y in range(height):
        for x in range(width):
            if x == 50 and y == 50:
                a = 999
                pass
            c = trace_ray(Ray(start=scene.camera.pos, dir=get_point(x, y, scene.camera)), scene, 0)
            pc = color.to_drawing_color(c)
            if x == 50 and y == 50:
                print('color=', pc)
            # print('c=',c)
            # print('pc=',pc)
            pixelOutput(x, y, pc)